connectors/src/connectors/notion/temporal/activities.ts

```diff
@@ -152,3 +249,201 @@ export async function getNotionAccessTokenActivity(
 
   return notionAccessToken;
 }
+
+async function shouldGarbageCollect(
+  dataSourceConfig: DataSourceConfig
+): Promise<boolean> {
+  const connector = await Connector.findOne({
+    where: {
+      type: "notion",
+      workspaceId: dataSourceConfig.workspaceId,
+      dataSourceName: dataSourceConfig.dataSourceName,
+    },
+    include: { model: NotionConnectorState, as: "notionConnectorState" },
+  });
+  if (!connector) {
+    throw new Error("Could not find connector");
+  }
+  const notionConnectorState = connector.notionConnectorState;
+  if (!notionConnectorState) {
+    throw new Error("Could not find notionConnectorState");
+  }
+
+  // If we have never finished a full sync, we should not garbage collect
+  const firstSuccessfulSyncTime = connector.firstSuccessfulSyncTime;
+  if (!firstSuccessfulSyncTime) {
+    return false;
+  }
+
+  const now = new Date().getTime();
+
+  // If we have never done a garbage collection, we should start one
+  // if it has been more than GARBAGE_COLLECTION_INTERVAL_HOURS since the first successful sync
+  if (!notionConnectorState.lastGarbageCollectionStartTime) {
+    return (
+      now - firstSuccessfulSyncTime.getTime() >=
+      GARBAGE_COLLECTION_INTERVAL_HOURS * 60 * 60 * 1000
+    );
+  }
+
+  const lastGarbageCollectionStartTime =
+    notionConnectorState.lastGarbageCollectionStartTime.getTime();
+
+  // If we have started a garbage collection, we should not start another one
+  if (!notionConnectorState.lastGarbageCollectionFinishTime) {
+    return false;
+  }
+  const lastGarbageCollectionFinishTime =
+    notionConnectorState.lastGarbageCollectionFinishTime.getTime();
+  if (lastGarbageCollectionStartTime > lastGarbageCollectionFinishTime) {
+    return false;
+  }
+
+  // if we garbage collected less than GARBAGE_COLLECTION_INTERVAL_HOURS ago, we should not start another one
+  if (
+    now - lastGarbageCollectionFinishTime <=
+    GARBAGE_COLLECTION_INTERVAL_HOURS * 60 * 60 * 1000
+  ) {
+    return false;
+  }
+
+  return true;
+}
+
+export async function saveStartGarbageCollectionActivity(
+  dataSourceConfig: DataSourceConfig
+) {
+  const transaction = await sequelize_conn.transaction();
+
+  try {
+    const connector = await Connector.findOne({
+      where: {
+        type: "notion",
+        workspaceId: dataSourceConfig.workspaceId,
+        dataSourceName: dataSourceConfig.dataSourceName,
+      },
+      include: { model: NotionConnectorState, as: "notionConnectorState" },
+    });
+
+    if (!connector) {
+      throw new Error("Could not find connector");
+    }
+
+    if (!connector.notionConnectorState) {
+      throw new Error("Could not find notionConnectorState");
+    }
+
+    await connector.notionConnectorState.update({
```

Comment:
I'm not a suppper fan of the result here, feels too magical to me. But this is a weakly held position, so happy to run with it.

Reply:
(Still weakly held)

Reply:
I'm fine with that, I don't mind the second query